src/websocket.js

```diff
@@ -0,0 +1,256 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import EventEmitter from 'events';
+import _WebSocket from 'isomorphic-ws';
+
+const WS_READYSTATE_READY = 1;
+
+
+/**
+ * Handles websocket connections and message transmission
+ *
+ * @class
+ * @name WS
+ */
+class WS extends EventEmitter {
+  constructor({
+    wsURL,
+    heartbeatInterval = 3000,
+    connectionTimeout = 5000,
+    retryConnectionInterval = 1000,
+    openConnectionTimeout = 20000,
+  } = {}) {
+    super();
+
+    // This is the class of the Websocket. It is used to open new WebSocket
+    // connections. We don't directly use it from import, so the unittests
+    // can replace it by a mock.
+    this.WebSocket = _WebSocket;
+
+    // This is the URL of the websocket.
+    this.wsURL = wsURL;
+
+    // Boolean to show when there is a websocket started with the server
+    this.started = false;
+
+    // Boolean to show when the websocket connection is working
+    this.connected = undefined;
+
+    // Store variable that is passed to Redux if ws is online
+    this.isOnline = undefined;
+
+    // Heartbeat interval in milliseconds.
+    this.heartbeatInterval = heartbeatInterval;
+
+    // Connection timeout.
+    this.connectionTimeout = connectionTimeout;
+
+    // Retry connection interval in milliseconds.
+    this.retryConnectionInterval = retryConnectionInterval;
+
+    // Open connection timeout.
+    this.openConnectionTimeout = openConnectionTimeout;
+
+    // Date of latest setup call. The setup is the way to open a new connection.
+    this.latestSetupDate = null;
+
+    // Date of latest ping.
+    this.latestPingDate = null;
+
+    // Latest round trip time measured by PING/PONG.
+    this.latestRTT = null;
+
+    // Timer used to detected when connection is down.
+    this.timeoutTimer = null;
+  }
+
+  /**
+   * Return websocket url to connect to.
+   **/
+  getWSServerURL() {
+    if (typeof this.wsURL === 'function') {
+      return this.wsURL();
+    } else {
+      return this.wsURL;
+    }
+  }
+
+  /**
+   * Start websocket object and its methods
+   */
+  setup() {
+    if (this.started) {
+      return;
+    }
+
+    const wsURL = this.getWSServerURL();
+    if (wsURL === null) {
+      // TODO Throw error?
+      return;
+    }
+
+    if (this.ws) {
+      const dt = (new Date() - this.latestSetupDate) / 1000;
+      if (dt < this.openConnectionTimeout) {
+        return;
+      }
+      this.ws.close();
```

Comment:
Yes, it will. When we create `this.ws` instance, we also set the callback for `this.ws.onClose`

```
this.ws.onclose = () => {
    this.onClose();
}
```

Reply:
Our `onClose` will try to restart the connection then:
```
setTimeout(() => {
      this.setup()
    }, this.retryConnectionInterval);
```

Don't we need to do `this.ws.onclose = () => {};`?

Reply:
It really makes no difference, the method `this.setup` is prepared to handle the cases where it's called multiple times. I will add this because maybe it's more correct semantically but it's not exactly a fix.